File: ui/src/services/server/api/link.ts
import { redisGet, redisSet } from "services/server/data/redis";
import { Link, LinkRequestDTO } from "types";
import { generateBase62String } from "utils/base62";
import { isValidUrl } from "utils/url";

const linkIDLength = (process.env.LINK_ID_LENGTH && parseInt(process.env.LINK_ID_LENGTH)) || 5;

export async function addLink(linkRequestDTO: LinkRequestDTO): Promise<Link> {
	if (!isValidUrl(linkRequestDTO.url)) {
		throw new Error("Invalid URL");
	}
	let id;
	let url;
	let tries = 0;
	do {
		id = generateBase62String(linkIDLength);
		// eslint-disable-next-line no-await-in-loop
		url = await redisGet(id);
	} while (url && ++tries < 10);
	if (url) {
		throw new Error("Maximum tries exceeded for link ID generation");
	}
	await redisSet(id, linkRequestDTO.url);
	return { id, url: linkRequestDTO.url };
}

export async function getLinkByID(id: string): Promise<Link> {
	const url = await redisGet(id);
	if (!url) {
		throw new Error("Link not found");
	}
	return { id, url };
}

File: services/api/link.go
package api

import (
	"errors"
	"slimlink/models"
	"slimlink/services/data"
	"slimlink/utils"
)

var linkIDLength int

func InitLinkApi(linkIDLength_ int) {
	linkIDLength = linkIDLength_
}

func AddLink(linkRequestDTO *models.LinkRequestDTO) (*models.Link, error) {
	if !utils.IsValidUrl(linkRequestDTO.Url) {
		return nil, errors.New("invalid URL")
	}
	var id string
	var url string
	var err error
	for tries := 0; tries < 10; tries++ {
		id = utils.GenerateBase62String(linkIDLength)
		url, err = data.RedisGet(id)
		if url == "" && err == nil {
			break
		}
	}
	if url != "" {
		return nil, errors.New("maximum tries exceeded for link ID generation")
	}
	err = data.RedisSet(id, linkRequestDTO.Url)
	if err != nil {
		return nil, err
	}
	return &models.Link{Id: id, Url: linkRequestDTO.Url}, nil
}

func GetLinkByID(id string) (*models.Link, error) {
	url, err := data.RedisGet(id)
	if err != nil {
		return nil, errors.New("failed to fetch data from Redis")
	}
	if url == "" {
		return nil, errors.New("link not found")
	}
	return &models.Link{Id: id, Url: url}, nil
}
